src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -34,138 +34,199 @@
-                     float uncertaintyOffset)
+HRVOAgent::HRVOAgent(HRVOSimulator *simulator, const Vector &position,
+                     float max_neighbor_dist, std::size_t maxNeighbors, float radius,
+                     const Vector &velocity, float maxAccel, AgentPath &path,
+                     float maxSpeed)
     : Agent(simulator, position, radius, velocity, velocity, maxSpeed, maxAccel, path),
       maxNeighbors_(maxNeighbors),
-      neighborDist_(neighborDist),
-      prefSpeed_(prefSpeed),
-      uncertaintyOffset_(uncertaintyOffset)
+      max_neighbor_dist(max_neighbor_dist),
+      prefSpeed_(max_speed_ * PREF_SPEED_SCALE),
+      // TODO (#2676): This config should be dependency injected and updated when a
+      // parameter is changed
+      obstacle_factory(TbotsProto::RobotNavigationObstacleConfig()),
+      ball_obstacle(std::nullopt)
 {
 }
 
-void HRVOAgent::computeNeighbors()
+void HRVOAgent::updatePrimitive(const TbotsProto::Primitive &new_primitive,
+                                const World &world)
+{
+    AgentPath path;
+    static_obstacles.clear();
+    ball_obstacle = std::nullopt;
+    if (new_primitive.has_move())
+    {
+        const auto &motion_control = new_primitive.move().motion_control();
+        float speed_at_dest        = new_primitive.move().final_speed_m_per_s();
+        float new_max_speed        = new_primitive.move().max_speed_m_per_s();
+        setMaxSpeed(new_max_speed);
+        setPreferredSpeed(new_max_speed * PREF_SPEED_SCALE);
+
+        // TODO (#2418): Update implementation of Primitive to support
+        // multiple path points and remove this check
+        CHECK(motion_control.path().points().size() >= 2)
+            << "Empty path: " << motion_control.path().points().size() << std::endl;
+        auto destination = motion_control.path().points().at(1);
+
+        // Max distance which the robot can travel in one time step + scaling
+        // TODO (#2370): This constant is calculated multiple times.
+        float path_radius = (max_speed_ * simulator_->getTimeStep()) / 2;
+        auto path_points  = {PathPoint(
+            Vector(destination.x_meters(), destination.y_meters()), speed_at_dest)};
+        path              = AgentPath(path_points, path_radius);
+
+        // Update static obstacles
+        std::set<TbotsProto::MotionConstraint> motion_constraints;
+        for (int constraint_int : motion_control.motion_constraints())
+        {
+            if (!TbotsProto::MotionConstraint_IsValid(constraint_int))
+            {
+                continue;
+            }
+
+            const auto constraint =
+                static_cast<TbotsProto::MotionConstraint>(constraint_int);
+            auto new_obstacles =
+                obstacle_factory.createFromMotionConstraint(constraint, world);
+            if (constraint == TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL)
+            {
+                ball_obstacle = new_obstacles[0];
```

Comment:
feels a bit weird to assume the ball obstacle will always be first

Reply:
I guess the assumption here is that 
```cpp
obstacle_factory.createFromMotionConstraint(TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL, world)
```
will return exactly one obstacle (i.e. half meter circle around the ball), which [is the case](https://github.com/UBC-Thunderbots/Software/blob/57f01c06ac16293a0d7048680a1306df5e92770f/src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp#L117). You should note that `new_obstacles` is not a list of all obstacles for all of the current motion constraints, it's a list of obstacles for just the current iteration's motion constraint.

Reply:
it may be cleaner to refactor it so that the obstacle factory returns a Map from MotionConstraint -> ObstaclePtr